### PR TITLE
fix:empty directory page after logged in for the first time

### LINKED ISSUE #47: Empty directory page after logged in for the first time
## Summary
When a newcomer logs in with GitHub authentication, the directory pool appears as an empty page.

## Description
Upon login, the page requests collaborator information for the logged-in user from the repository by calling `/repos/ubiquity/devpool-directory-private/collaborators/{username}`. This request receives a `404 NotFound` result, which is expected if the user is not a collaborator.

**Root Cause Analysis (from `jordan-ae`'s investigation):**
The issue is not related to caching. The error occurs within `fetch-issues-preview.ts`. The function attempts to fetch private issues first. For collaborators, the line `hasPrivateRepoAccess = await checkPrivateRepoAccess();` triggers a `403 error`. Consequently, the application bypasses the check for public repository issues and jumps to the catch clause, executing `handleRateLimit(octokit, error)`. This sets the `task` to `freshIssues`, which is an empty array due to the fetch failure. This empty task is then cached and displayed, resulting in a blank page.

This bug was introduced in PR #20.

## Expected Behavior
Even if the user is not a collaborator, the devpool listing should still be accessible on the listing page, similar to when the user is not logged in. The system should gracefully handle the `403 error` for private repo access and proceed to fetch public repo issues.

## Proposed Solution
The issue has been identified and a fix has been implemented in PR #48 (specifically `https://github.com/ubiquity/work.ubq.fi/pull/48/files#diff-a526a47f6685dd952321fd365cea2c75b28fb920d5d3f46f3dedb61d0ebed95bR27`). The solution involves adjusting the logic in `fetch-issues-preview.ts` to correctly handle the `403 error` for private repository access and ensure public issues are still fetched and displayed.

### PR DESCRIPTION
**Pull request by jordan-ae**
Resolves #47

Here's a concise pull request description for the changes you've made:

**Description:**
This pull request addresses an issue in the `fetchIssuePreviews` function where an error thrown by `checkPrivateRepoAccess` would prevent the population of the `freshIssues` array. The error handling logic was prematurely exiting the function, resulting in an empty array being returned.

Here's a rundown of what happens in the `fetchIssuePreview` function 

```js 
async function fetchIssuePreviews() {
  let freshIssues = [];
  try {
    // This will throw an error if the condition in checkPrivateRepoAccess's catch block is not met
    hasPrivateRepoAccess = await checkPrivateRepoAccess();
    // Code to populate freshIssues does not run, it moves directly to the catch block below...
  } catch (error) {
    // Error handling...
    // Since an error was thrown, freshIssues remains empty
  }
  // Returns tasks based on the empty freshIssues array
  return freshIssues.map(/* ... */);
}
```

Changes:
- Removed the `throw error` line in the `checkPrivateRepoAccess` function's `catch` block.
- This ensures that even if an error occurs, the `fetchIssuePreviews` function can continue to execute and populate `freshIssues` with issues from the public repository.

### CODE REVIEW

**Comment by jordan-ae** [comment]
[gentlementlegen](https://github.com/gentlementlegen) please can you review this? Added another cache related fix

**Comment by jordan-ae** [comment]
Hey @0x4007 I took a lot of time out to debug #49 I posted my findings concerning the root of the bug. I'd created a PR to address it. Took a quick break before committing and when I came back fix based on my findings was attached to another issue and closed this issue as completed! Does that mean I dont get the bounty and all the hours and effort I put into this were a waste. Despite me finding the bug and providing a fix?

**Comment by rndquu** [comment]
TBH I don't fully understand what to do with this PR

Timeline:
1. @jordan-ae self assigned this issue with deadline on 8 May and missed it
2. @gentlementlegen opened a [PR](https://github.com/ubiquity/work.ubq.fi/pull/48) fixing this issue on 10 May
3. @jordan-ae opened this PR on 13 May

@gentlementlegen Does this PR solve https://github.com/ubiquity/work.ubq.fi/issues/47 ? If https://github.com/ubiquity/work.ubq.fi/issues/47 is fixed and this PR doesn't solve the issue then we should close this PR.

**Comment by gentlementlegen** [comment]
@rndquu This PR (#47) coincidentally solved the issue because I needed [that specific part of the code](https://github.com/ubiquity/work.ubq.fi/pull/48/files#diff-a526a47f6685dd952321fd365cea2c75b28fb920d5d3f46f3dedb61d0ebed95bR27) the work in order to finish my task, which is why I ending up referencing it in my PR.

I think this PR should be closed without merge as another is worked on by @jordan-ae to fix a related problem. @jordan-ae please comfirm!

**Comment by jordan-ae** [comment]
@rndquu Well this PR contains a cache related fix that we'll potentially still have to do in the future. So it's up to you guys. I too am confused!

**Comment by rndquu** [comment]
> @rndquu Well this PR contains a cache related fix that we'll potentially still have to do in the future. So it's up to you guys. I too am confused!

I don't understand. Does this "cache related fix" connected with https://github.com/ubiquity/work.ubq.fi/issues/47? Or we should create a separate issue for it?

**Comment by jordan-ae** [comment]
Yes it's related to this issue

**Comment by Keyrxng** [comment]
- https://github.com/ubiquity/work.ubq.fi/blob/fed9bac6ab18303faf7b507491e5d8c74799e382/src/home/fetch-github/fetch-issues-preview.ts#L28
 
This `404: not a collaborator` line of code here should of been handling the user not being a collaborator if I'm not mistaken which was introduced months ago and the thrown error bubbled up through the `try{}catch{}`. Had a look over things to see if I should of been called to fix it if I broke it lmao. 

The [docs](https://docs.github.com/en/rest/collaborators/collaborators?apiVersion=2022-11-28#check-if-a-user-is-a-repository-collaborator) do state that it emits a 404 if the user is not a collaborator, and the 403 is `forbidden` so the issue must be `authToken` related, no?

**Comment by rndquu** [comment]
> * https://github.com/ubiquity/work.ubq.fi/blob/fed9bac6ab18303faf7b507491e5d8c74799e382/src/home/fetch-github/fetch-issues-preview.ts#L28
> 
> This `404: not a collaborator` line of code here should of been handling the user not being a collaborator if I'm not mistaken which was introduced months ago and the thrown error bubbled up through the `try{}catch{}`. Had a look over things to see if I should of been called to fix it if I broke it lmao.
> 
> The [docs](https://docs.github.com/en/rest/collaborators/collaborators?apiVersion=2022-11-28#check-if-a-user-is-a-repository-collaborator) do state that it emits a 404 if the user is not a collaborator, and the 403 is `forbidden` so the issue must be `authToken` related, no?

There is no 403 error in the screenshot [here](https://github.com/ubiquity/work.ubq.fi/issues/47#issue-2283071689) so the issue seems to be not `authToken` related

**Comment by rndquu** [comment]
@jordan-ae I can't reproduce the initial issue with empty page

[This](https://github.com/ubiquity/work.ubq.fi/pull/49/commits/6f0169bc03673ffcaa5a2f9d616ac8392b75122f) commit is meant to fix the issue while [this](https://github.com/ubiquity/work.ubq.fi/commit/2672cbf53f59098b03f1060d4817e9c831d98504) commit should contain the issue and display empty page on the 1st login.

When I'm on https://github.com/ubiquity/work.ubq.fi/commit/2672cbf53f59098b03f1060d4817e9c831d98504 how can I reproduce the issue with empty page? When I'm logged in for the 1st time with empty cache (i.e. local storage with saved github issues) I still see github issues while expected behaviour (based on issue [description](https://github.com/ubiquity/work.ubq.fi/issues/47)) is to get an empty page.

**Comment by Keyrxng** [comment]
> but for collaborators, the line hasPrivateRepoAccess = await checkPrivateRepoAccess(); triggers a 403 error

I was basing the 403 on this comment here and that before the auth flow got changed a while back I noticed that on first render after logging in octo wasn't provided the auth token, so I forced another login to fix it. Which looked like a page reload in all honestly but I thought it seemed similar, but I could be wrong

**Comment by jordan-ae** [comment]
> @jordan-ae I can't reproduce the initial issue with empty page

Hey @rndquu, the error initially wasn't easy to repro on development so I had to debug the issue directly on Production because that's where the bug was re-occuring easily.

Using Chrome DevTools, I tracked the event flow and identified an error causing the application to block collaborator issue loading. This error impacted everyone except core members, who have hasPrivateRepoAccess set to true, bypassing the problematic code.

My pull request #49 addressed the error and a caching issue that perpetuated it. 

After highlighting my [observations here](https://github.com/ubiquity/work.ubq.fi/issues/47#issuecomment-2106463470) @gentlementlegen sent in a fix for it. Wasn't sure what to do because I'd already prepared a PR for it so I sent it in anyway.

Since [#48](https://github.com/ubiquity/work.ubq.fi/pull/48) was merged first the error was fixed.

It's a bit frustrating on my part because it took me several days to track down the bug just for my observations to be used by someone else to resolve the issue. 

The fixes provided on this #49 & #48 for this issues are pretty close with #49 having an added cosmetic fix to it.

Was sending in my observations to try to get some opinions before sending in a PR. I didn't know it would end up costing me the issue. Guess I'd just have to take the L on this an let it go. 

This issue can pr can be closed since a PR fixing the bug has already been provided for it. Thanks

**Comment by rndquu** [comment]
> This error impacted everyone except core members

Yes, I was able to reproduce the issue. If you're logged in as a collaborator (i.e. not a core team member) then at https://github.com/ubiquity/work.ubq.fi/commit/2672cbf53f59098b03f1060d4817e9c831d98504 you're getting an empty list of issues.

The culprit is [this](https://github.com/ubiquity/work.ubq.fi/pull/49/files#diff-a526a47f6685dd952321fd365cea2c75b28fb920d5d3f46f3dedb61d0ebed95bL33) line.

This issue is solved by 2 PRs:
- https://github.com/ubiquity/work.ubq.fi/pull/48 (by [adding](https://github.com/ubiquity/work.ubq.fi/pull/48/files#diff-a526a47f6685dd952321fd365cea2c75b28fb920d5d3f46f3dedb61d0ebed95bR27) a more strict error check)
- https://github.com/ubiquity/work.ubq.fi/pull/49 (simply by [removing](https://github.com/ubiquity/work.ubq.fi/pull/49/files#diff-a526a47f6685dd952321fd365cea2c75b28fb920d5d3f46f3dedb61d0ebed95bL33) error throwing)

Timeline of events:
1. `@jordan-ae` self assigned this issue on 7 May with deadline on 8 May and missed it
2. `@gentlementlegen` opened a https://github.com/ubiquity/work.ubq.fi/pull/48 fixing this issue on 10 May (which was later merged)
3. `@jordan-ae` [posted](https://github.com/ubiquity/work.ubq.fi/issues/47#issuecomment-2106463470) a comment describing the root cause of an issue on 13 May
4. `@jordan-ae` opened this PR on 13 May (which is now closed)

So `@gentlementlegen` didn't have a change to see [this](https://github.com/ubiquity/work.ubq.fi/issues/47#issuecomment-2106463470) comment describing the root cause of an issue. 

TLDR:
- deadline was initially missed by `@jordan-ae`
- `@gentlementlegen` fixed the issue when it was unassigned + didn't get a chance to see the [comment](https://github.com/ubiquity/work.ubq.fi/issues/47#issuecomment-2106463470) describing the root cause

I'm inclined towards closing this PR and setting `@gentlementlegen` as an assignee of https://github.com/ubiquity/work.ubq.fi/issues/47 unless @0x4007 steps in.
